gpcontrib/arenadata_toolkit/src/track_files.c

```diff
@@ -0,0 +1,1448 @@
+				bloom_clear(&bloom_ctx.entry->bloom);
+				bloom_ctx.entry->work_version = track_bump_version(version);
+			}
+		}
+
+		bloom_set_release(&bloom_ctx);
+
+		/* initial snapshot shouldn't return drops */
+		if (!tf_get_global_state.bloom->is_set_all)
+		{
+			tf_get_global_state.drops = drops_track_move(MyDatabaseId);
+			tf_get_global_state.next_drop = list_head(tf_get_global_state.drops);
+		}
+
+		/*
+		 * Let's retrieve tracking information only once for the transaction.
+		 */
+		if (tf_get_global_state.schema_oids == NIL)
+			get_filters_from_guc();
+
+		if (tf_get_global_state.relstorages == 0 ||
+			tf_get_global_state.relkinds == 0 ||
+			tf_get_global_state.schema_oids == NIL)
+			ereport(ERROR,
+					(errmsg("Cannot get tracking configuration (schemas, relkinds, reltorage) for database %u", MyDatabaseId)));
+
+		MemoryContextSwitchTo(oldcontext);
+
+		oldcontext = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);
+
+		funcctx->tuple_desc = CreateTemplateTupleDesc(GET_TRACK_TUPDESC_LEN, false);
+		TupleDescInitEntry(funcctx->tuple_desc, Anum_track_relid + 1, "relid", OIDOID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, Anum_track_name + 1, "name", NAMEOID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, Anum_track_relfilenode + 1, "relfilenode", OIDOID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, Anum_track_size + 1, "size", INT8OID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, Anum_track_state + 1, "state", CHAROID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, Anum_track_gp_segment_id + 1, "gp_segment_id", INT4OID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, Anum_track_gp_segment_relnamespace + 1, "relnamespace", OIDOID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, Anum_track_gp_segment_relkind + 1, "relkind", CHAROID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, Anum_track_gp_segment_relstorage + 1, "relstorage", CHAROID, -1, 0);
+		funcctx->tuple_desc = BlessTupleDesc(funcctx->tuple_desc);
+
+		state = (tf_main_func_state_t *) palloc0(sizeof(tf_main_func_state_t));
+		funcctx->user_fctx = (void *) state;
+
+		state->pg_class_rel = heap_open(RelationRelationId, AccessShareLock);
+		state->scan = systable_beginscan(state->pg_class_rel, InvalidOid, false, NULL, 0, NULL);
+
+		MemoryContextSwitchTo(oldcontext);
+	}
+
+	funcctx = SRF_PERCALL_SETUP();
+	state = funcctx->user_fctx;
+
+	while (true)
+	{
+		Oid			filenode;
+		Oid			relnamespace;
+		char		relkind;
+		char		relstorage;
+		HeapTuple	pg_class_tuple;
+		Form_pg_class relp;
+		int64		size;
+
+		if (!state->scan)
+			break;
+
+		pg_class_tuple = systable_getnext(state->scan);
+
+		if (!HeapTupleIsValid(pg_class_tuple))
+		{
+			systable_endscan(state->scan);
+			heap_close(state->pg_class_rel, AccessShareLock);
+			state->scan = NULL;
+			state->pg_class_rel = NULL;
+			break;
+		}
+
+		datums[Anum_track_gp_segment_relkind] = heap_getattr(pg_class_tuple, Anum_pg_class_relkind, RelationGetDescr(state->pg_class_rel), &nulls[7]);
```

Comment:
Maybe you should use a macro instead of a magic number?
```suggestion
		datums[Anum_track_gp_segment_relkind] = heap_getattr(pg_class_tuple, Anum_pg_class_relkind, RelationGetDescr(state->pg_class_rel), &nulls[Anum_track_gp_segment_relkind]);
```